compiler/optimizations/remoteValueForwarding.cpp

```diff
@@ -208,10 +209,107 @@ static void updateTaskFunctions(Map<Symbol*, Vec<SymExpr*>*>& defMap,
       // Would need to flatten them if they are not already.
       INT_ASSERT(isGlobal(fn));
 
+      // Map from an aggregate type to a list of its fields that will be RVF'd
+      std::map<Symbol*, std::vector<Symbol*> > fieldsToRVF;
+      typedef std::map<Symbol*, std::vector<Symbol*> >::iterator argIter;
+
       // For each reference arg that is safe to dereference
       for_formals(arg, fn) {
         if (canForwardValue(defMap, useMap, syncSet, fn, arg)) {
           updateTaskArg(useMap, fn, arg);
+          AggregateType* argType = toAggregateType(arg->type);
+          if (argType && isRecord(argType)) {
+            for_fields(field, argType) {
+              if (isRecordWrappedType(field->getValType()) && field->isRef()) {
+                if (fieldsToRVF.find(arg) == fieldsToRVF.end()) {
+                  fieldsToRVF[arg] = std::vector<Symbol*>();
+                }
+                fieldsToRVF[arg].push_back(field);
+              }
+            }
+          }
+        }
+      }
+
+      if (fieldsToRVF.size() > 0) {
```

Comment:
I think it would be nice to put everything inside this conditional into a new function.